core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -92,7 +94,26 @@ public TableScan appendsAfter(long fromSnapshotId) {
     protected CloseableIterable<FileScanTask> planFiles(TableOperations ops, Snapshot snapshot, Expression rowFilter,
                                                         boolean ignoreResiduals, boolean caseSensitive,
                                                         boolean colStats) {
-      CloseableIterable<ManifestFile> filtered = filterManifests(manifests(), rowFilter, caseSensitive);
+      Map<Integer, PartitionSpec> specsById = table().specs();
+
+      LoadingCache<Integer, ManifestEvaluator> evalCache = Caffeine.newBuilder().build(specId -> {
+        PartitionSpec spec = specsById.get(specId);
+        PartitionSpec transformedSpec = transformSpec(fileSchema, spec, PARTITION_FIELD_PREFIX);
+        return ManifestEvaluator.forRowFilter(rowFilter, transformedSpec, caseSensitive);
+      });
+
+      CloseableIterable<ManifestFile> filtered = CloseableIterable.filter(
+          manifests(),
+          manifest -> {
+            PartitionSpec spec = specsById.get(manifest.partitionSpecId());
+
+            if (spec.fields().stream().anyMatch(f -> f.transform().equals(Transforms.alwaysNull()))) {
```

Comment:
This case is a bit ugly.  For V1 tables in the case of removing partition fields, it replaces the transform with a "void" transform.  What happens is that when we use this spec to evaluate a manifest file, it filters them all out as it searches for null value for this partition key, which seems incorrect.

Example, if we remove a field "data_bucket" from partition spec, then this code would check whether manifest files have null "data_bucket" values.

So hence skipping this check for now.

Reply:
Thought: haven't looked but maybe we could potentially push this logic down to ManifestEvaluator, return "ROWS_MIGHT_MATCH" match for any partition comparison involving  a partition field that has become VoidTransformation (removed in V1).  (I am thinking this case of querying removed partition field should only affect metadata tables primarily).

Reply:
Well, it looked wrong initially to filter out such manifests but I am no longer sure.
If we have a filter `partition.data_bucket = 1` and `data_bucket` is null for evolved tables, maybe it is correct to skip such manifests as technically the value for that partition value is null?

Reply:
Suppose I partitioned the table by `bucket(8, id)` and then switched to partitioning by `id` and have two files like this.

```
partition(data_bucket) = {1, null} - file written with bucket(8, id)
partition(id) = {null, 100} - file written with identity(id)
```

If I query `partition.data_bucket = 1`, it should be safe to skip the second file and manifest?

Reply:
Yea you may be right.  The difference is we are not skipping the manifest for V2 tables but are for V1, though maybe in the end they get filtered out anyway, I can add an end to end test for this.

Reply:
An end to end test would be great. I think it will produce the same output for V1 and V2 tables.

Reply:
Removed block and added test to show it doesnt matter